docs/docs/guides/03-typescript.md

```diff
@@ -2,4 +2,266 @@
+
+Definitions files provide a centralized location for types, making it easier to maintain and reuse them.
+
+Example:
+
+```typescript
+// service.d.ts
+interface CommonService {
+  [key: keyof any]: unknown;
+}
+
+// users.ts
+const usersService: CommonService = {
+  // ...
+};
+
+// books.ts
+const booksService: CommonService = {
+  // ...
+};
+```
+
+### c. Utils
+
+**For type utilities used across different packages**, it is mandatory to place them in their related module's definition file (.d.ts).
+
+- Strapi utils **must** be placed in `@strapi/strapi`'s types module
+- Packages' specific utils **must** be placed in their respective types' module
+
+Basically, utility modules contain reusable type definitions that can be imported wherever needed. This approach promotes code reusability and separation of concerns.
+
+Example:
+
+```typescript
+// core/uid.d.ts
+export type Admin = `admin::${string}`;
+
+// admin.ts
+import type { UID } from '@strapi/strapi';
+
+function getService(uid: UID.Admin): object {
+  // ...
+}
+```
+
+Choose the appropriate approach based on the size and complexity of your feature and the level of reusability of the types.
+
+## 2. Modules, Imports and Exports
+
+### a. Modules
+
+In TypeScript, modules can be used to organize and encapsulate related types and code. Modules provide a way to mimic namespaces and avoid polluting the global scope with numerous type declarations. By grouping related types within modules, it is possible to improve code organization and make it easier to navigate and understand the codebase.
+
+This is especially true when working with packages exports or utilities, where the exported components needs to be scoped.
+
+Example:
+
+```typescript
+// service.d.ts
+export type Admin = Record<string, any>;
+
+// uid.d.ts
+export type Admin = `admin::${string}`;
+
+// index.d.ts
+export * as Service from './service';
+export * as UID from './uid';
+
+// src/index.ts
+import type { Service, UID } from './index';
+
+type AdminService = Record<UID.Admin, Service.Admin>;
+```
+
+### b. Imports
+
+When importing **only types** from a module, `import type` must be used.
+
+The `import type` syntax imports only the types and eliminates the runtime impact of importing the actual values. This helps reduce the bundle size and improves performance by avoiding unnecessary code execution.
+
+Example:
+
+```typescript
+// ❌ don't
+
+import { Attribute, Common } from '@strapi/strapi';
+
+// ✅ do
+
+import type { Attribute, Common } from '@strapi/strapi';
+```
+
+When importing **both types and implementations** from a module:
+
+- `import type` must be used for types
+- `import` must be used for the rest
+
+Example:
+
+```typescript
+// ❌ don't
+
+import { User, findUser } from './user';
+
+const user: User = findUser();
+
+// ✅ do
+
+import type { User } from './user';
+import { findUser } from './user';
+
+const user: User = findUser();
+```
+
+### c. Exports
```

Comment:
I agree with what's said but that might just be too vague to be in the guideline 🤔 we probably need a specific list for this

Reply:
What kind of list are you thinking about?

It's basically: always use export {} except in a few situations like "`For example, when exporting a single primary entity or class from a module, or when working with modules that have a natural default export, such as a configuration object.`"

Reply:
Yes these are just examples :) we need clear rules with clear exceptions :)

Reply:
I've tried to make it clearer, not sure if this is what you meant. Waiting for your feedback

Reply:
[eff4e40](https://github.com/strapi/strapi/pull/17104/commits/eff4e403216736b51cf3ef9bc4fd281446adfe5e)

Reply:
Should be fine for now